python/felis/metadata.py

```diff
@@ -0,0 +1,509 @@
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+import logging
+from typing import IO, Any, Literal
+
+import sqlalchemy.schema as sqa_schema
+from sqlalchemy import (
+    CheckConstraint,
+    Column,
+    Constraint,
+    Engine,
+    ForeignKeyConstraint,
+    Index,
+    MetaData,
+    Numeric,
+    PrimaryKeyConstraint,
+    ResultProxy,
+    Table,
+    UniqueConstraint,
+    create_mock_engine,
+    make_url,
+    text,
+)
+from sqlalchemy.engine.mock import MockConnection
+from sqlalchemy.engine.url import URL
+from sqlalchemy.exc import SQLAlchemyError
+from sqlalchemy.types import TypeEngine
+
+from felis.datamodel import Schema
+from felis.db._variants import make_variant_dict
+
+from . import datamodel as dm
+from .db import sqltypes
+from .types import FelisType
+
+logger = logging.getLogger(__name__)
+
+
+class InsertDump:
+    """An Insert Dumper for SQL statements which supports writing messages
+    to stdout or a file.
+    """
+
+    def __init__(self, file: IO[str] | None = None) -> None:
+        """Initialize the insert dumper.
+
+        Parameters
+        ----------
+        file : `TextIOBase` or None, optional
+            The file to write the SQL statements to. If None, the statements
+            will be written to stdout.
+        """
+        self.file = file
+        self.dialect: Any | None = None
+
+    def dump(self, sql: Any, *multiparams: Any, **params: Any) -> None:
+        """Dump the SQL statement to a file or stdout.
+
+        Parameters
+        ----------
+        sql : `Any`
+            The SQL statement to dump.
+
+        multiparams : `Any`
+            The multiparams to use for the SQL statement.
+        params : `Any`
+            The params to use for the SQL statement.
+        """
+        compiled = sql.compile(dialect=self.dialect)
+        sql_str = str(compiled) + ";"
+        params_list = [compiled.params]
+        for params in params_list:
+            if not params:
+                print(sql_str, file=self.file)
+                continue
+            new_params = {}
+            for key, value in params.items():
+                if isinstance(value, str):
+                    new_params[key] = f"'{value}'"
+                elif value is None:
+                    new_params[key] = "null"
+                else:
+                    new_params[key] = value
+            print(sql_str % new_params, file=self.file)
```

Comment:
> Is it documented somewhere where the compiled sql string has `%` formatting in it?

Can you clarify what you want documented here? 

This is code that @andy-slac wrote - I can ask him if there are questions about it.

Reply:
I think my point relates back to not having any idea what the parameter for the method was (not Any) and then how somebody generates the string such that %-formatting is something that happens. A note in the docstring would be fine if it gave some context and pointed at some other docs.

Reply:
> I think my point relates back to not having any idea what the parameter for the method was (not Any) and then how somebody generates the string such that %-formatting is something that happens. A note in the docstring would be fine if it gave some context and pointed at some other docs.

This just makes sure that when parameters are provided to an executable statement, the resultant SQL is displayed correctly in the dump with the values inserted. I will make a note of this in the function.

Reply:
This was written before me, I just moved that code to a new location. That code does look a bit unsafe. It may work better if you just say 
```
    compiled = sql.compile(dialect=self.dialect, compile_kwargs={"literal_binds": True})
    print(str(compiled, file=...))
```

Reply:
> This was written before me, I just moved that code to a new location. That code does look a bit unsafe. It may work better if you just say 
> 
> ```
> 
>     compiled = sql.compile(dialect=self.dialect, compile_kwargs={"literal_binds": True})
> 
>     print(str(compiled, file=...))
> 
> ```

Thank you for the clarification. I'll try your suggested change.

Reply:
I've implemented the solution from @andy-slac and it seems to work well.

Reply:
@andy-slac and I will look at this on another ticket branch. There is probably an issue in the TAP visitor code preventing the suggested new code from working correctly.